File: scm-java/src/main/scheme/primitives/PrimitiveAesCbcDecrypt.java
package scheme.primitives;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.GeneralSecurityException;
import scheme.*;

public class PrimitiveAesCbcDecrypt extends Primitive {
    @Override
    public String name() { return "aes-cbc-decrypt"; }

    @Override
    public String info() {
        return "Syntax: (aes-cbc-decrypt key iv ciphertext)\n" +
               "Library: (scm crypto)\n" +
               "Description: Decrypts ciphertext using AES-CBC with PKCS7 padding. key must be 16, 24, or 32 bytes; iv must be 16 bytes. Returns a bytevector.\n" +
               "Example:\n" +
               "  (aes-cbc-decrypt key iv ciphertext) => #u8(...)";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 3, 3);
        byte[] key = Value.asBytevector(arguments[0]);
        byte[] iv = Value.asBytevector(arguments[1]);
        byte[] ciphertext = Value.asBytevector(arguments[2]);
        try {
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key, "AES"), new IvParameterSpec(iv));
            return cipher.doFinal(ciphertext);
        } catch (GeneralSecurityException e) {
            throw new SchemeError(pos, "aes-cbc-decrypt: " + e.getMessage());
        }
    }
}

File: scm-csharp/scheme/primitives/PrimitiveAesCbcDecrypt.cs
using System.Security.Cryptography;

namespace scheme;

public class PrimitiveAesCbcDecrypt : Primitive
{
    public override string Name() => "aes-cbc-decrypt";

    public override string Info() =>
        "Syntax: (aes-cbc-decrypt key iv ciphertext)\n" +
        "Library: (scm crypto)\n" +
        "Description: Decrypts ciphertext using AES-CBC with PKCS7 padding. key must be 16, 24, or 32 bytes; iv must be 16 bytes. Returns a bytevector.\n" +
        "Example:\n" +
        "  (aes-cbc-decrypt key iv ciphertext) => #u8(...)";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 3, 3);
        byte[] key = Value.AsBytevector(arguments[0]);
        byte[] iv = Value.AsBytevector(arguments[1]);
        byte[] ciphertext = Value.AsBytevector(arguments[2]);
        using var aes = Aes.Create();
        aes.Mode = CipherMode.CBC;
        aes.Padding = PaddingMode.PKCS7;
        aes.Key = key;
        aes.IV = iv;
        using var decryptor = aes.CreateDecryptor();
        return decryptor.TransformFinalBlock(ciphertext, 0, ciphertext.Length);
    }
}
